### ISSUE TITLE
Kong 1.0.3 upgrade to 1.1.0rc2 with C* migrations problems

### ISSUE THREAD

**Issue by jeremyjpj0916** [Kong/kong#4420]
### Summary

So I am running Kong 1.0.3 nodes, trying to upgrade our dev to 1.1.0rc2 to ensure nothing has broken in latest iteration. I first spin up a new temporary pod on 1.1.0rc2 against the existing 1.0.3 nodes C* datastore, running the kong migrations up command against it, this pod reports that there are migrations needing finished on the existing keyspace and exits out. So I go to check my existing 1.0.3 nodes running against this datastore to check if there were old migrations to finish, they all say no extra kong migrations finish needed from their perspective.

![checkKong1 0 3PodMigrationsStatus](https://user-images.githubusercontent.com/31913027/54806505-8a20a280-4c50-11e9-91e0-388d758260e2.png)

I checked the keyspace looking for the new tags table I expected with latest Kong but I did not see the table present in the keyspace, which if it had been present I would have assumed somehow my temporary pod executed the migrations up command twice somehow and second call threw the error but it seems to not be the case. Current present tables after that error was thrown: 

![KongCurrentTables](https://user-images.githubusercontent.com/31913027/54806784-775a9d80-4c51-11e9-9f01-4c4bdec4f0cd.png)

I also know this is not ideal flow in the new model - start a temp pod, run the up command to update db, then update the other pods to latest kong version in rolling deploy fashion as per discussion here: https://discuss.konghq.com/t/kong-1-1-0rc1-available-for-testing/2873/5?u=jeremyjpj0916 , but my approach still should have worked I think in this situation. Ideal impl now being I start up perma new pods on new version run check the need for migrations, run migrations up on those pods, tear down old pods, then run migrations finish from the new pods.

Other ideas I have are maybe go ahead and try to start up the 1.1.0rc2 nodes against the datastore anyways even though I don't see the new "tags" table. I don't have high hopes in that working though as I would expect the tags table to be present at this moment. Don't wanna try that until I raised the initial issue here first.

Under additional details you can see below logs of the migrations up from the newer node when it came up as a temp pod and executed the migrations up against existing keyspace.

Also attaching my keyspace here for reference:

[kong_dev.zip](https://github.com/Kong/kong/files/2995355/kong_dev.zip)

### Steps To Reproduce

1. Be running Kong 1.0.3 kong node, startup a kong 1.1.0rc2 node and execute kong migrations up. 
2. If it behaves like mine, the 1.1.0rc2 pod will fail on the migrations up call saying there are migrations pending and to run kong migrations finish. At which point idk what to do as I expected the migrations up to make new tables but keep db in flexible state between Kong versions 1.0.3 and 1.1.0rc2 and then as I take down my old pods and bring up the new I run finish on the new pods. 

### Additional Details & Logs

- Kong version 1.1.0rc2
- Kong debug-level startup logs (`$ kong start --vv`)
```
2019/03/22 06:48:00 [verbose] Kong: 1.1.0rc2
2019/03/22 06:48:00 [debug] ngx_lua: 10013
2019/03/22 06:48:00 [debug] nginx: 1013006
2019/03/22 06:48:00 [debug] Lua: LuaJIT 2.1.0-beta3
2019/03/22 06:48:00 [verbose] no config file found at /etc/kong/kong.conf
2019/03/22 06:48:00 [verbose] no config file found at /etc/kong.conf
2019/03/22 06:48:00 [verbose] no config file, skip loading
2019/03/22 06:48:00 [debug] reading environment variables
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_USERNAME ENV found with "*****"
2019/03/22 06:48:00 [debug] KONG_PG_USER ENV found with ""
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_DATA_CENTERS ENV found with "DC2:3,DC1:3"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_PASSWORD ENV found with "******"
2019/03/22 06:48:00 [debug] KONG_PG_HOST ENV found with ""
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_SSL ENV found with "on"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_PORT ENV found with "9042"
2019/03/22 06:48:00 [debug] KONG_DATABASE ENV found with "cassandra"
2019/03/22 06:48:00 [debug] KONG_PG_DATABASE ENV found with ""
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_LOCAL_DATACENTER ENV found with "DC1"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_CONSISTENCY ENV found with "LOCAL_QUORUM"
2019/03/22 06:48:00 [debug] KONG_PG_SSL_VERIFY ENV found with ""
2019/03/22 06:48:00 [debug] KONG_PG_SSL ENV found with "off"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_REPL_STRATEGY ENV found with "NetworkTopologyStrategy"
2019/03/22 06:48:00 [debug] KONG_PG_PASSWORD ENV found with "******"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_SCHEMA_CONSENSUS_TIMEOUT ENV found with "120000"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_CONTACT_POINTS ENV found with "server00466"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_TIMEOUT ENV found with "20000"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_SSL_VERIFY ENV found with "on"
2019/03/22 06:48:00 [debug] KONG_LUA_SSL_VERIFY_DEPTH ENV found with "3"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_REPL_FACTOR ENV found with "3"
2019/03/22 06:48:00 [debug] KONG_PG_PORT ENV found with ""
2019/03/22 06:48:00 [debug] KONG_LUA_SSL_TRUSTED_CERTIFICATE ENV found with "/usr/local/kong/ssl/kongcert.pem"
2019/03/22 06:48:00 [warn] You are using Cassandra but your 'db_update_propagation' setting is set to '0' (default). Due to the distributed nature of Cassandra, you should increase this value.
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_LB_POLICY ENV found with "RequestDCAwareRoundRobin"
2019/03/22 06:48:00 [debug] KONG_CASSANDRA_KEYSPACE ENV found with "kong_dev"
2019/03/22 06:48:00 [debug] KONG_NGINX_DAEMON ENV found with "off"
2019/03/22 06:48:00 [debug] admin_access_log = "logs/admin_access.log"
2019/03/22 06:48:00 [debug] admin_error_log = "logs/error.log"
2019/03/22 06:48:00 [debug] admin_listen = {"127.0.0.1:8001","127.0.0.1:8444 ssl"}
2019/03/22 06:48:00 [debug] anonymous_reports = true
2019/03/22 06:48:00 [debug] cassandra_consistency = "LOCAL_QUORUM"
2019/03/22 06:48:00 [debug] cassandra_contact_points = {"server00466"}
2019/03/22 06:48:00 [debug] cassandra_data_centers = {"DC2:3","DC1:3"}
2019/03/22 06:48:00 [debug] cassandra_keyspace = "kong_dev"
2019/03/22 06:48:00 [debug] cassandra_lb_policy = "RequestDCAwareRoundRobin"
2019/03/22 06:48:00 [debug] cassandra_local_datacenter = "DC1"
2019/03/22 06:48:00 [debug] cassandra_password = "******"
2019/03/22 06:48:00 [debug] cassandra_port = 9042
2019/03/22 06:48:00 [debug] cassandra_repl_factor = 3
2019/03/22 06:48:00 [debug] cassandra_repl_strategy = "NetworkTopologyStrategy"
2019/03/22 06:48:00 [debug] cassandra_schema_consensus_timeout = 120000
2019/03/22 06:48:00 [debug] cassandra_ssl = true
2019/03/22 06:48:00 [debug] cassandra_ssl_verify = true
2019/03/22 06:48:00 [debug] cassandra_timeout = 20000
2019/03/22 06:48:00 [debug] cassandra_username = "*******"
2019/03/22 06:48:00 [debug] client_body_buffer_size = "8k"
2019/03/22 06:48:00 [debug] client_max_body_size = "0"
2019/03/22 06:48:00 [debug] client_ssl = false
2019/03/22 06:48:00 [debug] database = "cassandra"
2019/03/22 06:48:00 [debug] db_cache_ttl = 0
2019/03/22 06:48:00 [debug] db_resurrect_ttl = 30
2019/03/22 06:48:00 [debug] db_update_frequency = 5
2019/03/22 06:48:00 [debug] db_update_propagation = 0
2019/03/22 06:48:00 [debug] dns_error_ttl = 1
2019/03/22 06:48:00 [debug] dns_hostsfile = "/etc/hosts"
2019/03/22 06:48:00 [debug] dns_no_sync = false
2019/03/22 06:48:00 [debug] dns_not_found_ttl = 30
2019/03/22 06:48:00 [debug] dns_order = {"LAST","SRV","A","CNAME"}
2019/03/22 06:48:00 [debug] dns_resolver = {}
2019/03/22 06:48:00 [debug] dns_stale_ttl = 4
2019/03/22 06:48:00 [debug] error_default_type = "text/plain"
2019/03/22 06:48:00 [debug] headers = {"server_tokens","latency_tokens"}
2019/03/22 06:48:00 [debug] log_level = "notice"
2019/03/22 06:48:00 [debug] lua_package_cpath = ""
2019/03/22 06:48:00 [debug] lua_package_path = "./?.lua;./?/init.lua;"
2019/03/22 06:48:00 [debug] lua_socket_pool_size = 30
2019/03/22 06:48:00 [debug] lua_ssl_trusted_certificate = "/usr/local/kong/ssl/kongcert.pem"
2019/03/22 06:48:00 [debug] lua_ssl_verify_depth = 3
2019/03/22 06:48:00 [debug] mem_cache_size = "128m"
2019/03/22 06:48:00 [debug] nginx_admin_directives = {}
2019/03/22 06:48:00 [debug] nginx_daemon = "off"
2019/03/22 06:48:00 [debug] nginx_http_directives = {}
2019/03/22 06:48:00 [debug] nginx_optimizations = true
2019/03/22 06:48:00 [debug] nginx_proxy_directives = {}
2019/03/22 06:48:00 [debug] nginx_sproxy_directives = {}
2019/03/22 06:48:00 [debug] nginx_stream_directives = {}
2019/03/22 06:48:00 [debug] nginx_user = "nobody nobody"
2019/03/22 06:48:00 [debug] nginx_worker_processes = "auto"
2019/03/22 06:48:00 [debug] origins = {}
2019/03/22 06:48:00 [debug] pg_ssl = false
2019/03/22 06:48:00 [debug] pg_ssl_verify = false
2019/03/22 06:48:00 [debug] pg_timeout = 5000
2019/03/22 06:48:00 [debug] plugins = {"bundled"}
2019/03/22 06:48:00 [debug] prefix = "/usr/local/kong/"
2019/03/22 06:48:00 [debug] proxy_access_log = "logs/access.log"
2019/03/22 06:48:00 [debug] proxy_error_log = "logs/error.log"
2019/03/22 06:48:00 [debug] proxy_listen = {"0.0.0.0:8000","0.0.0.0:8443 ssl"}
2019/03/22 06:48:00 [debug] real_ip_header = "X-Real-IP"
2019/03/22 06:48:00 [debug] real_ip_recursive = "off"
2019/03/22 06:48:00 [debug] ssl_cipher_suite = "modern"
2019/03/22 06:48:00 [debug] ssl_ciphers = "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256"
Error: 
/usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:165: [Cassandra error] cluster_mutex callback threw an error: /usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:139: database has pending migrations; run 'kong migrations finish'
stack traceback:
	[C]: in function 'error'
	/usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:139: in function </usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:135>
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/kong/db/init.lua:345: in function </usr/local/share/lua/5.1/kong/db/init.lua:295>
	[C]: in function 'pcall'
	/usr/local/share/lua/5.1/kong/concurrency.lua:42: in function 'cluster_mutex'
	/usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:135: in function 'up'
	/usr/local/share/lua/5.1/kong/cmd/migrations.lua:139: in function 'cmd_exec'
	/usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:88>
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:45>
	/usr/local/bin/kong:7: in function 'file_gen'
	init_worker_by_lua:48: in function <init_worker_by_lua:46>
	[C]: in function 'xpcall'
	init_worker_by_lua:55: in function <init_worker_by_lua:53>
stack traceback:
	[C]: in function 'error'
	/usr/local/share/lua/5.1/kong/cmd/utils/migrations.lua:165: in function 'up'
	/usr/local/share/lua/5.1/kong/cmd/migrations.lua:139: in function 'cmd_exec'
	/usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:88>
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/kong/cmd/init.lua:88: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:45>
	/usr/local/bin/kong:7: in function 'file_gen'
	init_worker_by_lua:48: in function <init_worker_by_lua:46>
	[C]: in function 'xpcall'
	init_worker_by_lua:55: in function <init_worker_by_lua:53>
2019/03/22 06:48:00 [debug] stream_listen = {"off"}
2019/03/22 06:48:00 [debug] trusted_ips = {}
2019/03/22 06:48:00 [debug] upstream_keepalive = 60
2019/03/22 06:48:00 [verbose] prefix in use: /usr/local/kong
2019/03/22 06:48:00 [debug] resolved Cassandra contact point 'server00466' to: 10.xx.xxx.xx
2019/03/22 06:48:00 [debug] loading subsystems migrations...
2019/03/22 06:48:00 [verbose] retrieving keyspace schema state...
2019/03/22 06:48:01 [verbose] schema state retrieved
2019/03/22 06:48:01 [debug] loading subsystems migrations...
2019/03/22 06:48:01 [verbose] retrieving keyspace schema state...
2019/03/22 06:48:01 [verbose] schema state retrieved
```

**Comment by jeremyjpj0916** [comment]
Figure I would drop this in too if its ringing any alarms as well as I was digging:
![KeyspaceSchemaMeta](https://user-images.githubusercontent.com/31913027/54807104-8c83fc00-4c52-11e9-920c-c5a49544d961.png) , I saw one row is not null, could that be the culprit and if so why does my 1.0.3 nodes say db is up to date and migrations finish has nothing to do?

**Comment by jeremyjpj0916** [comment]
Also for more ammo here is our config via ENV variables as I think problem is related to not including bundled plugins that have a migration at some version:

```
- name: KONG_PLUGINS
          value: >-
            zipkin,kong-error-log,kong-oidc-implicit-token,kong-response-size-limiting,request-transformer,kong-service-virtualization,kong-cluster-drain,kong-upstream-jwt,kong-splunk-log,kong-spec-expose,kong-path-based-routing,kong-oidc-multi-idp,correlation-id,oauth2,statsd,jwt,rate-limiting,acl,request-size-limiting,request-termination,cors
        - name: KONG_DNS_ORDER
          value: 'LAST,SRV,A,CNAME'
        - name: KONG_CASSANDRA_CONTACT_POINTS
          value: 'server00466,server00467,server00468,server00469,server00470,server00471'
        - name: KONG_LOG_LEVEL
          value: notice
        - name: KONG_PROXY_ACCESS_LOG
          value: 'off'
        - name: KONG_ADMIN_ACCESS_LOG
          value: 'off'
        - name: KONG_PROXY_ERROR_LOG
          value: /dev/stderr
        - name: KONG_ADMIN_ERROR_LOG
          value: /dev/stderr
        - name: KONG_ANONYMOUS_REPORTS
          value: 'off'
        - name: KONG_PROXY_LISTEN
          value: '0.0.0.0:8000, 0.0.0.0:8443 ssl http2'
        - name: KONG_ADMIN_LISTEN
          value: '0.0.0.0:8001'
        - name: KONG_MEM_CACHE_SIZE
          value: 1024m
        - name: KONG_SSL_CERT
          value: /usr/local/kong/ssl/kongcert.crt
        - name: KONG_SSL_CERT_KEY
          value: /usr/local/kong/ssl/kongprivatekey.key
        - name: KONG_SSL_CERT_DER
          value: /usr/local/kong/ssl/kongcertder.der
        - name: KONG_CLIENT_SSL
          value: 'off'
        - name: KONG_UPSTREAM_KEEPALIVE
          value: '0'
        - name: KONG_HEADERS
          value: latency_tokens
        - name: KONG_REAL_IP_HEADER
          value: X-Forwarded-For
        - name: KONG_REAL_IP_RECURSIVE
          value: 'on'
        - name: KONG_CLIENT_MAX_BODY_SIZE
          value: 1000m
        - name: KONG_CLIENT_BODY_BUFFER_SIZE
          value: 20m
        - name: KONG_ERROR_DEFAULT_TYPE
          value: text/plain
        - name: KONG_DATABASE
          value: cassandra
        - name: KONG_PG_SSL
          value: 'off'
        - name: KONG_CASSANDRA_PORT
          value: '9042'
        - name: KONG_CASSANDRA_KEYSPACE
          value: kong_dev
        - name: KONG_CASSANDRA_TIMEOUT
          value: '5000'
        - name: KONG_CASSANDRA_SSL
          value: 'on'
        - name: KONG_CASSANDRA_SSL_VERIFY
          value: 'on'
        - name: KONG_CASSANDRA_USERNAME
          value: dbusername
        - name: KONG_CASSANDRA_PASSWORD
          value: dbpassword
        - name: KONG_CASSANDRA_CONSISTENCY
          value: LOCAL_QUORUM
        - name: KONG_CASSANDRA_LB_POLICY
          value: RequestDCAwareRoundRobin
        - name: KONG_CASSANDRA_LOCAL_DATACENTER
          value: DC1
        - name: KONG_DB_UPDATE_FREQUENCY
          value: '5'
        - name: KONG_DB_UPDATE_PROPAGATION
          value: '5'
        - name: KONG_DB_CACHE_TTL
          value: '0'
        - name: KONG_DNS_HOSTSFILE
          value: /etc/hosts
        - name: KONG_DNS_STALE_TTL
          value: '4'
        - name: KONG_DNS_NOT_FOUND_TTL
          value: '30'
        - name: KONG_DNS_ERROR_TTL
          value: '1'
        - name: KONG_DNS_NO_SYNC
          value: 'off'
        - name: KONG_LUA_SSL_TRUSTED_CERTIFICATE
          value: /usr/local/kong/ssl/kongcert.pem
        - name: KONG_LUA_SSL_VERIFY_DEPTH
          value: '3'
        - name: KONG_LUA_SOCKET_POOL_SIZE
          value: '30'
        - name: KONG_TRUSTED_IPS
          value: '0.0.0.0/0,::/0'
```

**Comment by thibaultcha** [comment]
@jeremyjpj0916 It seems to me that when you migrated your <= 0.14 database during the 1.0 upgrade, `plugins = bundled` was in effect (it is the default), which made the response-ratelimiting migrations run. The 0.15 to 1.0 migration of this plugin has a teardown step, which runs during `kong migrations finish` (https://github.com/Kong/kong/blob/master/kong/plugins/response-ratelimiting/migrations/002_15_to_10.lua). Now before you ran `finish` you may have started setting the `plugins` property as you described above (i.e. excluding the response-ratelimiting migrations), which left as pending migration behind for this plugin. It then seems like you ran your 1.1.0rc2 migrations _without_ specifying the `plugins` property, which is then defaulted to `bundled`, which means that the leftover pending migration is considered a blocker. But your 1.0.3 nodes don't even see it, since the plugin is disabled for them.

Now to fix this issue, I suggest running `kong migrations finish` from a 1.0.3 node in which you specify `plugins=bundled` (you could do this without touching any configuration files or environment with `KONG_PLUGINS=bundled kong migrations finish`).

Do you see it as a possibility that the initial 0.14 to 1.0 migration ran without changing the `plugins` property, and that your `finish` step had a more restrictive list of plugins at the time?

**Comment by thibaultcha** [comment]
I was able to reproduce this behavior like so:

```
$ export KONG_DATABASE=cassandra
$ cd kong/

$ git checkout 0.14.1
   # install
$ kong migrations up
...
66 migrations ran
waiting for Cassandra schema consensus (10000ms timeout)...
Cassandra schema consensus: reached
$ git checkout 1.0.3
    # install
$ kong migrations up
upgrading from 0.14, bootstrapping database...
migrating core on keyspace 'kong'...
core migrated up to: 000_base (executed)
core migrated up to: 001_14_to_15 (pending)
core migrated up to: 002_15_to_1 (pending)
migrating oauth2 on keyspace 'kong'...
oauth2 migrated up to: 000_base_oauth2 (executed)
oauth2 migrated up to: 001_14_to_15 (pending)
oauth2 migrated up to: 002_15_to_10 (pending)
migrating acl on keyspace 'kong'...
acl migrated up to: 000_base_acl (executed)
acl migrated up to: 001_14_to_15 (pending)
migrating jwt on keyspace 'kong'...
jwt migrated up to: 000_base_jwt (executed)
jwt migrated up to: 001_14_to_15 (executed)
migrating basic-auth on keyspace 'kong'...
basic-auth migrated up to: 000_base_basic_auth (executed)
basic-auth migrated up to: 001_14_to_15 (executed)
migrating key-auth on keyspace 'kong'...
key-auth migrated up to: 000_base_key_auth (executed)
key-auth migrated up to: 001_14_to_15 (executed)
migrating rate-limiting on keyspace 'kong'...
rate-limiting migrated up to: 000_base_rate_limiting (executed)
rate-limiting migrated up to: 001_14_to_15 (executed)
rate-limiting migrated up to: 002_15_to_10 (pending)
migrating hmac-auth on keyspace 'kong'...
hmac-auth migrated up to: 000_base_hmac_auth (executed)
hmac-auth migrated up to: 001_14_to_15 (executed)
migrating response-ratelimiting on keyspace 'kong'...
response-ratelimiting migrated up to: 000_base_response_rate_limiting (executed)
response-ratelimiting migrated up to: 001_14_to_15 (executed)
response-ratelimiting migrated up to: 002_15_to_10 (pending)
22 migrations processed
15 executed
7 pending
$ export KONG_PLUGINS=acl,oauth2,rate-limiting
$ kong migrations finish
migrating core on keyspace 'kong'...
core migrated up to: 001_14_to_15 (executed)
core migrated up to: 002_15_to_1 (executed)
migrating rate-limiting on keyspace 'kong'...
rate-limiting migrated up to: 002_15_to_10 (executed)
migrating acl on keyspace 'kong'...
acl migrated up to: 001_14_to_15 (executed)
migrating oauth2 on keyspace 'kong'...
oauth2 migrated up to: 001_14_to_15 (executed)
oauth2 migrated up to: 002_15_to_10 (executed)
6 migrations processed
6 executed
```

At this point, the `schema_meta` table shows that the response-ratelimiting plugin hasn't been fully migrated:
```
 key         | subsystem             | executed                                                   | last_executed | pending
-------------+-----------------------+------------------------------------------------------------+---------------+------------------
 schema_meta |                   acl |                           {'000_base_acl', '001_14_to_15'} |  001_14_to_15 |             null
 schema_meta |            basic-auth |                    {'000_base_basic_auth', '001_14_to_15'} |  001_14_to_15 |             null
 schema_meta |                  core |                {'000_base', '001_14_to_15', '002_15_to_1'} |   002_15_to_1 |             null
 schema_meta |             hmac-auth |                     {'000_base_hmac_auth', '001_14_to_15'} |  001_14_to_15 |             null
 schema_meta |                   jwt |                           {'000_base_jwt', '001_14_to_15'} |  001_14_to_15 |             null
 schema_meta |              key-auth |                      {'000_base_key_auth', '001_14_to_15'} |  001_14_to_15 |             null
 schema_meta |                oauth2 |        {'000_base_oauth2', '001_14_to_15', '002_15_to_10'} |  002_15_to_10 |             null
 schema_meta |         rate-limiting | {'000_base_rate_limiting', '001_14_to_15', '002_15_to_10'} |  002_15_to_10 |             null
 schema_meta | response-ratelimiting |        {'000_base_response_rate_limiting', '001_14_to_15'} |  001_14_to_15 | {'002_15_to_10'}
``` 

Then, running the 1.1.0rc2 migrations:

```
$ git checkout 1.1.0
    # install
$ kong start
database has pending migrations:
response-ratelimiting: 002_15_to_10
database has new migrations available:
core: 003_100_to_110
run 'kong migrations up' to proceed
Error: ./kong/cmd/start.lua:50: nginx: [warn] [lua] cluster.lua:497: refresh(): [lua-cassandra] foun contact point with '0.0.0.0' as rpc_address, using '127.0.0.1' to contact it instead. If this is inorrect you should avoid using '0.0.0.0' in rpc_address
nginx: [error] init_by_lua error: ./kong/init.lua:340: new migrations available; run 'kong migration list'
...
$ kong migrations up
Error: [Cassandra error] cluster_mutex callback threw an error: ./kong/cmd/utils/migrations.lua:139: database has pending migrations; run 'kong migrations finish'
...
$ export KONG_PLUGINS=bundled
$ kong migrations finish
migrating response-ratelimiting on keyspace 'kong'...
response-ratelimiting migrated up to: 002_15_to_10 (executed)
1 migration processed
1 executed
$ kong migrations up
migrating core on keyspace 'kong'...
core migrated up to: 003_100_to_110 (executed)
1 migration processed
1 executed
$ unset KONG_PLUGINS
$ kong start
Kong started
```

The lesson here seems that we should be doing a better job at informing users that:

1. The `plugins` property has an impact on which migrations get to run (both "up" and "finish" steps)
2. Some migrations subsystems (in this case, plugins) may be blocking a migration attempt because they were not specified by the user (basically have migrations bypass the `plugins` property, sort of)
3. Make `kong migrations list` output reflects pending migrations from disabled subsystems (i.e. disabled plugins)

Thanks for the report @jeremyjpj0916!

**Comment by jeremyjpj0916** [comment]
@thibaultcha excellent analysis as per usual, after reading just a few sentences of what you were dishing out it dawned on me. In the past when we first started working with kong the "migration" was always a OpenShift job, a temporary kong pod comes up, runs the kong migrations command and goes back down. This migrations pod that would come up and tear down in the past we never cared about the config value, We just saw it as a way to bring the db and standard tables up to date. The complication comes I I suppose as soon as that plugins field became available in Kong and defaulted to bundled if not explicitly set. Since our real pods never run bundled trying to save on perf and only specific plugins we use, and our temporary "migrations" pod we just set the bare minimum env variables needed to connect to the db there lies the discrepancy 😄 and why bundled was set on the migrations step. Then we would go back to run kong migrations finish from our actual kong pods after starting up on the new version(with our explicit plugins set) thus leaving off plugins that the migrations kong pod "bundled" had established changes for.

So yes it seems the near term "fix" on our end would be to bring the current 1.03 pods up with env bundled included to make sure all the core plugins get their migrations ran at this moment with a kong migrations finish, then I could drop the bundled and likely upgrade to 1.1 without issue ensuring I now factor in setting the plugins field during first migrations call in my flow at the moment. I know we certainly need to move away from handling migrations like this, but as a temp fix I can make my migrations job plugins field mirror what our actual Kong environments run. Then plugins our actual kong pods don't use won't get stuck like this again.


And lessons I agree 3. would have helped me connect the dots here in output. 


Thanks! I think we both got some good takeaways here to grow on.
